teps/draft-20200716-task-results-in-final-tasks.md

```diff
@@ -0,0 +1,279 @@
+With adding support for `Task` `Results`, we are introducing dependencies to finally `Tasks` but that
+does not change scheduling of finally tasks. All final tasks will still be executed in parallel after
+all non-final tasks finishes.
+
+**Q. What happens to `Pipeline` when the dependent task `boskos-acquire` either failed or not executed and
+the task result `leased-resource` is not initialized?**
+
+**A.** The finally `Task` `boskos-release` is attempted but declared failure since the task result
+`leased-resource` is not initialized. Pipeline controller logs the reason of this failure as invalid result
+reference `InvalidTaskResultReference`. `Pipeline` continues executing rest of the final tasks and exits with failure.
+
+**Q. What happens when the dependent `Task` succeeds but the `Task` `Result` is empty?**
+
+**A.** It could be possible that the task author intentionally chooses to leave the `Task` `Result` empty
+(creating an empty file at `/tekton/results/`). `Pipeline` leaves that param value empty and executes
+the final task. Empty `Task` `Result` is considered a valid value and results in successful variable
+substitution. The error handling of such empty `Task` `Result` is left to the final `Task` author.
+
+**Q. Can a final `Task` produce `Task` `Results`?**
+
+**A.** Nope, this proposal does not enable final tasks to produce `Task` `Results`.
+
+## Advantages
+
+* No API changes.
+
+* Finally contract does not break and all final tasks are attempted.
+
+* Consistent across non-final and final tasks.
+
+* Finally tasks do not need to explicitly check if the Task they depend on ran or not
+
+* This proposal is extensible with a param validation proposal using Schema -
+[#1393](https://github.com/tektoncd/pipeline/issues/1393#issuecomment-558833526) by reading default from
+`Task` specification.
+
+> `PipelineRun` reads parameter defaults from `Pipeline` specification but in case of
+parameters associated with `Task` `Results`, the default values will be used from the `Task` since
+`Pipeline` has no direct mapping between `Pipeline` parameter and `Task` parameter.
+
+```yaml
+# Pipeline
+spec:
+  params:
+    - name: owner-name
+      default: "the-best-owner"
+  tasks:
+   - name: boskos-acquire
+     taskRef:
+       name: boskos-acquire
+     params:
+       - name: owner-name
+         value: $(params.owner-name)
+   - name: do-stuff-with-resource
+     taskRef:
+       name: do-stuff-with-resource
+     params:
+       - name: resource-name
+         value:  $(tasks.boskos-acquire.results.leased-resource)
+ finally:
+   - name: boskos-release
+     taskRef:
+       name: boskos-release
+     params:
+       - name: leased-resource
+         value:  $(tasks.boskos-acquire.results.leased-resource)
+---
+
+# Task
+name: boskos-release
+spec:
+  params:
+    - name: resource-name
+      default: "best-default-ever"
+      schema:
+        type: string
+        minLen: 1
```

Comment:
i'm a bit confused by this example: if the default value of the param will be used, I don't think the schema validation will come into play at all?

Reply:
I can drop the `schema` from here for now, but the param validation must be done on the defaults as well once we introduce `schema`. And such param validation happens before scheduling any tasks in the pipeline for execution. Its considered more of a syntax validation i.e. part of compilation (if it make sense 😉 )